bqskit/passes/search/generators/seed.py

```diff
@@ -0,0 +1,127 @@
+"""This module implements the SeedLayerGenerator class."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from bqskit.ir.circuit import Circuit
+from bqskit.passes.search.generator import LayerGenerator
+from bqskit.passes.search.generators.simple import SimpleLayerGenerator
+from bqskit.qis.state.state import StateVector
+from bqskit.qis.unitary.unitarymatrix import UnitaryMatrix
+from bqskit.utils.typing import is_integer
+_logger = logging.getLogger(__name__)
+
+
+class SeedLayerGenerator(LayerGenerator):
+    """Layer Generator for search that starts from a seed."""
+
+    def __init__(
+        self,
+        seed: Circuit,
+        forward_generator: LayerGenerator = SimpleLayerGenerator(),
+        num_removed: int = 1,
+    ) -> None:
+        """
+        Construct a SimpleLayerGenerator.
+
+        Args:
+            seed (Circuit): The seed to start from.
+
+            forward_generator (Gate): A generator used to grow the circuit.
+
+            num_removed (int): The number of gates removed from the circuit
+                in each backwards branch.
+        """
+        if not isinstance(seed, Circuit):
+            raise TypeError(f'Expected Circuit for seed, got {type(seed)}')
+
+        if not isinstance(forward_generator, LayerGenerator):
+            raise TypeError(
+                'Expected LayerGenerator for forward_generator'
+                f', got {type(forward_generator)}.',
+            )
+
+        if not is_integer(num_removed):
+            raise TypeError(
+                f'Expected integer for num_removed, got {type(num_removed)}.',
+            )
+
+        self.seed = seed
+        self.forward_generator = forward_generator
+        self.num_removed = num_removed
+
+    def gen_initial_layer(
+        self,
+        target: UnitaryMatrix | StateVector,
+        data: dict[str, Any],
+    ) -> Circuit:
+        """
+        Generate the initial layer, see LayerGenerator for more.
+
+        Raises:
+            ValueError: If `target` has a size or radix mismatch with
+                `self.seed`.
+        """
+
+        if not isinstance(target, (UnitaryMatrix, StateVector)):
+            raise TypeError(
+                'Expected unitary or state, got %s.' % type(target),
+            )
+
+        if target.dim != self.seed.dim:
+            raise ValueError('Seed dimension mismatch with target.')
+
+        data['seed_seen_before'] = [self.hash_structure(self.seed)]
```

Comment:
Perhaps we should make this a set and implement `__hash__` on `Circuit`?

Reply:
Yeah, I'll make it a set.

I am hesitant to add a `__hash__` to `Circuit` since it is a mutable object. I use hashes here because the `Circuit`s here are representing circuit structures and are not going to change.

Reply:
Good point! In that case I think this is fine.